### ISSUE TITLE
Handler fails to restart service cleanly (task works fine)

### ISSUE THREAD

**Issue by wfaler** [ansible/ansible-modules-core#3018]
#### ISSUE TYPE

Bug Report
##### COMPONENT NAME

service module
##### ANSIBLE VERSION

2.0.0.2
##### SUMMARY

I have a handler that looks as follows on Ubuntu 14.04.1 (running OS X 10.11 and Ansible 2.0.0.2 on the controlling machine).

```

---
- name: restart unicorn_server
  action: service name=unicorn_server state=restarted
```

The restart block of the service script looks as follows:

``` restart|reload|upgrade)
    echo "Upgrading $APP_NAME"
    if start-stop-daemon --stop --signal USR2 --quiet --oknodo --pidfile $PID; then
      sleep 3
      if start-stop-daemon --stop --signal QUIT --quiet --oknodo --pidfile $PID.oldbin; then
        exit 0
      else
        exit 1
      fi
    else
      exit 1
    fi
    ;;
```

Running the script manually with `sudo service unicorn_server restart` runs the service as expected.
Also running the service as a task in a playbook works fine.
However, when the handler is called, somehow the process is killed, but never starts again.

I am somewhat puzzled as to why the handler manages to only kill the service, whereas both manual and task based execution of the service manages to send the correct set of signals to the process without killing it.

**Comment by MichaelBaydoun** [comment]
Is the handler being called with sudo?

**Comment by wfaler** [comment]
It should be, the playbook itself is run as sudo, with the exception of a few tasks.

**Comment by mvaled** [comment]
Just is case it's relevant.  In my case the error shown is like:

```
RUNNING HANDLER [pyramidapp : reload nginx] ************************************
fatal: [paiement.com]: FAILED! => {"changed": false, "failed": true, "msg": "reload: Job is not running: nginx\n"}
```

Notice the trailing `\n` in the service name.

My handler is simply:

```
  - name: reload nginx
    service: >-
      name=nginx
      state=reloaded

```

The handler is placed inside a block with `become: yes`.  I'm running ansible from a github clone, currently at:
- ansible: fac3bac72e524aee36728a7dd026d3429710be61
- core module: 13f4b40c15d995351ed77e05a4253488b2ef4608

**Comment by asmodehn** [comment]
FYI I just had a similar issue (handler for service restart not working, but task or direct module call is fine), also on on Ubuntu 14.04, with

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

I eventually tracked it down to a permission issue.
From memory (take it with a grain of salt), my hierarchy was : 

```
playbook -> myrole+become -> dependency1_role+become.
                          -> dependency2_role+handler.
```

My handler didn't get the privilege escalation, somehow...
I reviewed all my `become:yes` everywhere and cleaned it up to :

```
playbook+become -> myrole -> dependency1_role.
                          -> dependency2_role+handler.
```

and it started to work...

Not sure if that become was the actual problem or not, but that would be my first guess.
So just in case, you might want to double check if it is a permission issue or not when running service module inside a handler...

**Comment by asmodehn** [comment]
Actually my problem was not fixed by changing the `become` setting. So I made a separate issue for it.

**Comment by MichaelBaydoun** [comment]
Have you tried running the playbook with -vvvv and did what was the output?

**Comment by ansibot** [comment]
@grossws, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@grossws, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by darrenob** [comment]
I have the same issue. Even when trying through an ad-hoc command:

```
$ ansible -i env app1 -m service -a"name=myservice state=restarted" -bK
SUDO password:
app1 | SUCCESS => {
    "changed": true,
    "name": "myservice",
    "state": "started"
```

On the host, it's been killed:

```
[root@app1 ~]# service myservice status
Not running (process dead but pidfile exists)
```

This only happens for 2 specific services, one of them shiny-server, the other a custom init script for an in-house java process. All other services can be managed ok by Ansible. Both of the bad ones restart ok manually on the host:

```
[root@app1 ~]# service myservice restart
Stopped [20293]
Started [20324]
```

This not only happens with the service module. It also happens when I try the command module and strangely, the output shows to have restarted the service (after it was killed earlier), complete with a new PID:

```
$ ansible -i env app1 -m command -a"/etc/init.d/myservice restart" -bK
SUDO password:
app1 | SUCCESS | rc=0 >>
Not running (process dead but pidfile exists)
Started [20696]
```

But on the server it's still not running. That PID is non existent.

```
[root@app1 ~]# ps aux | grep 20696
root     21448  0.0  0.0 103304   868 pts/2    S+   19:16   0:00 grep 21326
```

The full init script is here if that helps: https://gist.github.com/darrenob/32239a925121c7a24adbc58196ee89b5

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mgodwin** [comment]
Hate to +1 this, but recently was bit by this one as well.  Forgot the `become: yes` directive in my handler and couldn't bounce `apache2`.  Scary because it 'appeared' to work but definitely didn't bounce the service.  Did confirm that running it as a task w/o the correct privileges does blow up.  Also adding `become: yes` does fix the problem, but am a little uneasy about it now that I got a false positive.

**Comment by darrenob** [comment]
@ mgodwin Thanks for the response. Unfortunately that doesn't work for me. The service doesn't bounce even when `become: yes`

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
This issue was migrated to https://github.com/ansible/ansible/issues/29899
